### PR TITLE
Allow to resolve `authenticationOptions` based on authenticator name or display name

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Proposed changes in this pull request
Currently in adaptive scripts we use authenticator friendly names to resolve `authenticationOptions` parameter. However this raises troubles when the authenticator display names get changed.
We should use authenticator names instead of friendly names in adaptive scripts. Friendly names should be expect to change.
To preserve backward compatibility, we are introducing a system flag.

Added to the existing `AdaptiveAuth` section. By default this will be enabled.
```sh
<AuthenticatorNameInAuthConfig>
    <Enable>true</Enable>
</AuthenticatorNameInAuthConfig>
```

### Related issue
- https://github.com/wso2/product-is/issues/14843

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 1]
```diff
@@ -207,6 +207,9 @@ public class FrameworkUtils {
     private static final String HASH_CHAR = "#";
     private static final String HASH_CHAR_ENCODED = "%23";
     private static final String QUESTION_MARK = "?";
+    public static final String AUTHENTICATOR_NAME_IN_AUTH_CONFIG = "AdaptiveAuth.AuthenticatorNameInAuthConfig.Enable";
```
Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:210]: do we need this as public?

Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:210]: can move this to https://github.com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java#L561

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 [hunk 2]
```diff
@@ -2884,6 +2884,10 @@
             {% endfor %}
         </ChoreoDomains>
         {% endif %}
+
+        <AuthenticatorNameInAuthConfig>
```
Comment (dewniMW) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:2888]: better to add a small comment explaining the purpose of this config. 
also let's get this config added to identity.xml

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 3]
```diff
@@ -3296,4 +3299,24 @@ public static String buildCallerPathRedirectURL(String callerPath, Authenticatio
...
+        if (authenticatorNameInAuthConfigPreference == null) {
+            String serverConfig = IdentityUtil.getProperty(AUTHENTICATOR_NAME_IN_AUTH_CONFIG);
+            if (StringUtils.isNotBlank(serverConfig)) {
+                authenticatorNameInAuthConfigPreference = Boolean.parseBoolean(serverConfig);
+            } else {
+                authenticatorNameInAuthConfigPreference = false;
+            }
+        }
+
+        return authenticatorNameInAuthConfigPreference;
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 (hunk 2 above)
Comment (ThaminduDilshan) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:2889]: Fixed

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml [hunk 4]
```diff
@@ -2102,6 +2102,11 @@
 
         <!--Timeout in milliseconds for the waiting external calls-->
         <LongWaitTimeout>10000</LongWaitTimeout>
+
+        <!--Whether to use authenticator name or display name in authenticationOptions-->
```
Comment (dewniMW) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:2106]:
```suggestion
        <!--Whether to use authenticator name in authenticationOptions. If set to false, display name will be used in authenticationOptions.-->
```

**Review by dewniMW — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 5]
```diff
@@ -517,6 +517,8 @@ public static class JSAttributes {
         public static final String FORCE_AUTH_PARAM = "forceAuth";
         public static final String SUBJECT_IDENTIFIER_PARAM = "markAsSubjectIdentifierStep";
         public static final String SUBJECT_ATTRIBUTE_PARAM = "markAsSubjectAttributeStep";
+        public static final String AUTHENTICATOR_NAME_IN_AUTH_CONFIG =
+                "AdaptiveAuth.AuthenticatorNameInAuthConfig.Enable";
```
Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:521]: use the AdaptiveAuthentication inner class

Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java]: Fixed

**Review by ShanChathusanda93 — APPROVED** [review]

**Pull request merged**